src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1645,6 +1697,23 @@ void MacroAssembler::movptr(Register Rd, address addr, int32_t &offset) {
   offset = imm64 & 0x3f;
 }
 
+void MacroAssembler::movptr2(Register Rd, uint64_t addr, int32_t &offset, Register tmp) {
+  assert_different_registers(Rd, tmp, noreg);
+
+  uint32_t upper18 = (addr >> 30ull);
```

Comment:
literal suffix `ull` could be removed?
And `uint32_t upper18 = (addr >> 30ull);` + `lui(tmp, upper18 << 12);` could be replaced with `lui(tmp, addr >> 18);`?

Reply:
**ull** is there to say should be a 64-bit logical shift, appereant not doing it's jobs.
The reason it's done in two steps is first we get the values, then we adjust them for our implementation of the instructions.

Our lui() shifts down the immediate value so we must shift it up before.
While when we patch lui we use the value without shifting:
```
  unsigned int upper18 = (addr >> 30ull);
  Assembler::patch(instruction_address + (NativeInstruction::instruction_size * 0), 31, 12, (upper18 & 0xfffff)); 

```
So it can be replaced but then we are missing the calculation why it's correct,

Reply:
Thanks for explanation.